src/bin/pg_autoctl/fsm_transition.c

```diff
@@ -412,40 +413,84 @@ prepare_replication(Keeper *keeper, bool other_node_missing_is_ok)
 	 */
 	if (!config->monitorDisabled)
 	{
-		if (!monitor_get_other_node(monitor, config->nodename, pgSetup->pgport,
-									&(keeper->otherNode)))
+		char *host = config->nodename;
+		int port = pgSetup->pgport;
+
+		if (!monitor_get_other_nodes(monitor, host, port,
+									 otherNodeState,
+									 &(keeper->otherNodes)))
 		{
-			if (other_node_missing_is_ok)
+			/* errors have already been logged */
+			return false;
+		}
+
+		if (keeper->otherNodes.count == 0)
+		{
+			if (otherNodeState == ANY_STATE)
 			{
-				log_debug("There's no other node for %s:%d",
-						  config->nodename, pgSetup->pgport);
+				log_warn("There's no other node for %s:%d", host, port);
 			}
 			else
 			{
-				log_error("There's no other node for %s:%d",
-						  config->nodename, pgSetup->pgport);
+				/*
+				 * Should we warn about it really? it might be a replication
+				 * setting change that will impact synchronous_standby_names
+				 * and that's all.
+				 */
+				log_warn("There's no other node in state \"%s\" "
+						 "for node %s:%d",
+						 NodeStateToString(otherNodeState),
+						 host, port);
 			}
-			return other_node_missing_is_ok;
 		}
 	}
 
-	if (!primary_add_standby_to_hba(postgres,
-									keeper->otherNode.host,
-									config->replication_password))
+	/*
+	 * Should we fail somewhere in this loop, we return false and fail the
+	 * whole transition. The transition is going to be tried again, and we are
+	 * going to try and add HBA entries and create replication slots again.
+	 * Both operations succeed when their target entry already exists.
+	 *
+	 * Note that primary_create_replication_slot() is idempotent thanks to
+	 * first dropping the target replication slot, then creating it again. We
+	 * might want to avoid drop/create noise on those servers that we managed
+	 * to process in the previous loop.
```

Comment:
Yes, dropping the slot means standby that are a little behind might not be able to catch up anymore.

Reply:
Good point. I implemented “if not exists” behaviour instead here.